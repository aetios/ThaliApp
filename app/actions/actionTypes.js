export const NAVIGATE = 'NAVIGATE';
export const BACK = 'BACK';
export const OPENDRAWER = 'OPENDRAWER';
export const CALENDARRETREIVED = 'CALENDARRETREIVED';
export const CALENDARERROR = 'CALENDARERROR';
export const RESETLOGINSTATE = 'RESETLOGINSTATE';
export const WELCOME = 'WELCOME';
