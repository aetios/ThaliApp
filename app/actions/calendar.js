/**
 * Created by aucke on 3/8/17.
 */
