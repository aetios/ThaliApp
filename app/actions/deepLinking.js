export const DEEPLINK = 'DEEPLINKING_DEEPLINK';

export function deepLink(url, stayInApp = true) {
  return {
    type: DEEPLINK,
    payload: { url, stayInApp },
  };
}
