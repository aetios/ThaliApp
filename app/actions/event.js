export const EVENT = 'EVENT_EVENT';
export const FETCHING = 'EVENT_FETCHING';
export const SUCCESS = 'EVENT_SUCCESS';
export const FAILURE = 'EVENT_FAILURE';

export function event(pk, token) {
  return {
    type: EVENT,
    payload: { pk, token },
  };
}

export function success(eventData, eventRegistrations) {
  return {
    type: SUCCESS,
    payload: { eventData, eventRegistrations },
  };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}

export function failure() {
  return {
    type: FAILURE,
  };
}
