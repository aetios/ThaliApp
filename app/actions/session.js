export const LOGIN = 'SESSION_LOGIN';
export const INIT = 'SESSION_INIT';
export const SUCCESS = 'SESSION_SUCCESS';
export const LOGOUT = 'SESSION_LOGOUT';
export const TOKEN_INVALID = 'SESSION_TOKEN_INVALID';
export const PROFILE = 'SESSION_PROFILE';
export const PROFILE_SUCCESS = 'SESSION_PROFILE_SUCCESS';

export function success(username, token) {
  return { type: SUCCESS, payload: { username, token } };
}

export function login(user, pass) {
  return { type: LOGIN, payload: { user, pass } };
}

export function init() {
  return { type: INIT };
}

export function logout() {
  return { type: LOGOUT };
}

export function tokenInvalid() {
  return { type: TOKEN_INVALID };
}

export function profile(token) {
  return { type: PROFILE, payload: { token } };
}

export function profileSuccess(displayName, photo) {
  return { type: PROFILE_SUCCESS, payload: { displayName, photo } };
}
