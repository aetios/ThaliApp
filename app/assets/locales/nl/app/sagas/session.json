{
  "Login failed": "Inloggen mislukt",
  "Logout successful": "Succesvol uitgelogd",
  "Login successful": "Succesvol ingelogd"
}
