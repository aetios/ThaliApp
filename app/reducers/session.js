import { defaultProfileImage } from '../utils/url';

import * as sessionActions from '../actions/session';

const initialState = {
  token: '',
  username: '',
  displayName: '',
  photo: defaultProfileImage,
};

export default function session(state = initialState, action = {}) {
  switch (action.type) {
    case sessionActions.SUCCESS:
      return {
        ...state,
        username: action.payload.username,
        token: action.payload.token,
      };
    case sessionActions.PROFILE_SUCCESS:
      return {
        ...state,
        displayName: action.payload.displayName,
        photo: action.payload.photo,
      };
    case sessionActions.TOKEN_INVALID:
    case sessionActions.LOGOUT:
      return initialState;
    default:
      return state;
  }
}
