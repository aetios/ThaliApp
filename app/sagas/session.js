import {
  call, put, takeEvery, select,
} from 'redux-saga/effects';
import { delay } from 'redux-saga';
import { AsyncStorage } from 'react-native';
import Snackbar from 'react-native-snackbar';
import { Sentry } from 'react-native-sentry';

import { apiRequest } from '../utils/url';
import * as sessionActions from '../actions/session';
import * as pushNotificationsActions from '../actions/pushNotifications';
import { tokenSelector } from '../selectors/session';

export const USERNAMEKEY = '@MyStore:username';
export const TOKENKEY = '@MyStore:token';
export const DISPLAYNAMEKEY = '@MyStore:displayName';
export const PHOTOKEY = '@MyStore:photo';
export const PUSHCATEGORYKEY = '@MyStore:pushCategories';

const pairsToObject = (obj, pair) => {
  const obj2 = { ...obj };
  obj2[pair[0]] = pair[1];
  return obj2;
};

function* init() {
  try {
    const result = yield call([AsyncStorage, 'multiGet'], [
      USERNAMEKEY, TOKENKEY, DISPLAYNAMEKEY, PHOTOKEY, PUSHCATEGORYKEY,
    ]);
    const values = result.reduce(pairsToObject, {});

    const username = values[USERNAMEKEY];
    const token = values[TOKENKEY];
    const displayName = values[DISPLAYNAMEKEY];
    const photo = values[PHOTOKEY];
    const pushCategories = JSON.parse(values[PUSHCATEGORYKEY]);

    if (username !== null && token !== null) {
      yield put(sessionActions.signedIn(username, token));
      yield put(sessionActions.setUserInfo(displayName, photo));
      yield put(sessionActions.fetchUserInfo());
      yield put(pushNotificationsActions.register(pushCategories));
    } else {
      yield put(sessionActions.tokenInvalid());
    }
  } catch (e) {
    Sentry.captureException(e);
  }
}

function* signIn(action) {
  const { user, pass } = action.payload;

  const data = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username: user,
      password: pass,
    }),
  };

  const currentTimestamp = Date.now();

  try {
    const response = yield call(apiRequest, 'token-auth', data);
    const { token } = response;

    yield call(AsyncStorage.multiSet, [
      [USERNAMEKEY, user],
      [TOKENKEY, token],
    ]);
    yield put(sessionActions.signedIn(user, token));
    yield put(sessionActions.fetchUserInfo());
    yield put(pushNotificationsActions.register());
    Snackbar.show({ title: 'Login successful' });
  } catch (e) {
    // Delay failure to make sure animation is finished
    const now = Date.now();
    if (now - currentTimestamp < 150) {
      yield call(delay, now - currentTimestamp);
    }

    yield put(sessionActions.tokenInvalid());
    Sentry.captureException(e);
    Snackbar.show({ title: 'Login failed' });
  }
}

function* clearUserInfo() {
  yield call(AsyncStorage.clear);
  yield put(pushNotificationsActions.invalidate());
}

function* signOut() {
  yield call(clearUserInfo);
  Snackbar.show({ title: 'Logout successful' });
}

function* signedIn({ payload }) {
  const { username } = payload;
  yield call(Sentry.setUserContext, { username });
}

function* userInfo() {
  const token = yield select(tokenSelector);

  const data = {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Token ${token}`,
    },
  };

  try {
    const userProfile = yield call(apiRequest, 'members/me', data);

    yield call(AsyncStorage.multiSet, [
      [DISPLAYNAMEKEY, userProfile.display_name],
      [PHOTOKEY, userProfile.avatar.medium],
    ]);
    yield put(sessionActions.setUserInfo(userProfile.display_name, userProfile.avatar.medium));
  } catch (error) {
    Sentry.captureException(error);
  }
}

function* sessionSaga() {
  yield takeEvery(sessionActions.INIT, init);
  yield takeEvery(sessionActions.SIGN_IN, signIn);
  yield takeEvery(sessionActions.SIGN_OUT, signOut);
  yield takeEvery(sessionActions.SIGNED_IN, signedIn);
  yield takeEvery(sessionActions.FETCH_USER_INFO, userInfo);
  yield takeEvery(sessionActions.TOKEN_INVALID, clearUserInfo);
}

export default sessionSaga;
