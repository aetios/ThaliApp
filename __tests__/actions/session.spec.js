import * as actions from '../../app/actions/session';

describe('session actions', () => {
  it('should expose the session actions', () => {
    expect(actions.INIT).toEqual('SESSION_INIT');
    expect(actions.SUCCESS).toEqual('SESSION_SUCCESS');
    expect(actions.LOGIN).toEqual('SESSION_LOGIN');
    expect(actions.TOKEN_INVALID).toEqual('SESSION_TOKEN_INVALID');
    expect(actions.LOGOUT).toEqual('SESSION_LOGOUT');
    expect(actions.PROFILE).toEqual('SESSION_PROFILE');
    expect(actions.PROFILE_SUCCESS).toEqual('SESSION_PROFILE_SUCCESS');
  });

  it('should create an action to init the session', () => {
    expect(actions.init()).toMatchSnapshot();
  });

  it('should create an action to notify invalid token', () => {
    expect(actions.tokenInvalid()).toMatchSnapshot();
  });

  it('should create an action to log the user in', () => {
    expect(actions.login('username', 'password')).toMatchSnapshot();
  });

  it('should create an action for a successful login', () => {
    expect(actions.success('username', 'token')).toMatchSnapshot();
  });

  it('should create an action to log the user out', () => {
    expect(actions.logout()).toMatchSnapshot();
  });

  it('should create an action to load the user profile', () => {
    expect(actions.profile('token')).toMatchSnapshot();
  });

  it('should create an action for a successful user profile load', () => {
    expect(actions.profileSuccess('displayName', 'photo')).toMatchSnapshot();
  });
});