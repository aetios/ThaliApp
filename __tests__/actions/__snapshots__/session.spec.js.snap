// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session actions should create an action for a successful login 1`] = `
Object {
  "payload": Object {
    "token": "token",
    "username": "username",
  },
  "type": "SESSION_SUCCESS",
}
`;

exports[`session actions should create an action for a successful user profile load 1`] = `
Object {
  "payload": Object {
    "displayName": "displayName",
    "photo": "photo",
  },
  "type": "SESSION_PROFILE_SUCCESS",
}
`;

exports[`session actions should create an action to init the session 1`] = `
Object {
  "type": "SESSION_INIT",
}
`;

exports[`session actions should create an action to load the user profile 1`] = `
Object {
  "payload": Object {
    "token": "token",
  },
  "type": "SESSION_PROFILE",
}
`;

exports[`session actions should create an action to log the user in 1`] = `
Object {
  "payload": Object {
    "pass": "password",
    "user": "username",
  },
  "type": "SESSION_LOGIN",
}
`;

exports[`session actions should create an action to log the user out 1`] = `
Object {
  "type": "SESSION_LOGOUT",
}
`;

exports[`session actions should create an action to notify invalid token 1`] = `
Object {
  "type": "SESSION_TOKEN_INVALID",
}
`;
